File: advent_of_code/typescript/src/2015/day01.ts
import { getInputChallenge } from "../services/adventofcode";

const dict = {
  "(": 1,
  ")": -1,
};

function run() {
  partOne();
  partTwo();
}

const partOne = () => {
  getInputChallenge({ year: 2015, day: 1 })
    .then((res) => {
      let sum = 0;

      for (let char of res.data) {
        sum += dict[char as keyof typeof dict];
      }

      console.log({ sum });
    })
    .catch((err) => {
      console.error({ err });
    });
};

const partTwo = () => {
  getInputChallenge({ year: 2015, day: 1 })
    .then((res) => {
      const text = res.data;
      let sum = 0;
      let pos = 0;
      for (let i = 0; i < text.length; i++) {
        sum += dict[text[i] as keyof typeof dict];

        if (sum === -1) {
          pos = i + 1;
          break;
        }
      }
      console.log({ sum, pos });
    })
    .catch((err) => console.error({ err }));
};

run();

File: advent_of_code/rust/aoc/src/years/mod_2015/day_01.rs
use aoc::get_input_example;

pub fn run() {
    let data = get_input_example(2015, 1);
    println!(
        "2015, day_01, part_one:{}, part_two:{}",
        part_one(&data),
        part_two(&data)
    )
}
fn part_one(data: &str) -> i32 {
    let mut sum = 0;

    for c in data.chars() {
        match c {
            '(' => sum += 1,
            ')' => sum -= 1,
            _ => sum += 0,
        }
    }

    sum
}

fn part_two(data: &str) -> usize{
    let mut sum = 0;
    let mut pos = 0;
    for (i, c) in data.chars().enumerate() {
        match c {
            '(' => sum += 1,
            ')' => sum -= 1,
            _ => sum += 0,
        }
        if sum == -1 {
            pos = i + 1;
            break;
        }
    }
    pos
}
